File: leetcode/0905_sort_array_by_parity.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <algorithm>
#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
    vector<int> sortArrayByParity(vector<int>& A) {
        int i = 0, j = A.size() - 1;
        while (i < j) {
            if (A[i] % 2 == 0) {
                i++;
            } else if (A[j] % 2 == 1) {
                j--;
            } else {
                swap(A[i++], A[j--]);
            }
        }

        return A;
    }
};

int main() {
    auto solution = Solution();

    vector<int> A = {3, 1, 2, 4};

    vector<int> expected = {4, 2, 1, 3};
    vector<int> result = solution.sortArrayByParity(A);

    assert(expected == result);
}

File: leetcode/0905_sort_array_by_parity.py
# -*- coding: utf-8 -*-


class Solution:
    def sortArrayByParity(self, A):
        i, j = 0, len(A) - 1
        while i < j:
            if A[i] % 2 == 0:
                i += 1
            elif A[j] % 2 == 1:
                j -= 1
            elif A[i] % 2 == 1 and A[j] % 2 == 0:
                A[i], A[j] = A[j], A[i]
                i += 1
                j -= 1

        return A


if __name__ == "__main__":
    solution = Solution()

    assert [4, 2, 1, 3] == solution.sortArrayByParity([3, 1, 2, 4])
